File: basis/src/main/java/com/basis/ui/BaseFragment.java
package com.basis.ui;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.IdRes;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.basis.UIStack;
import com.basis.mvp.BasePresenter;
import com.basis.mvp.IBaseView;
import com.basis.net.LoadTag;
import com.kit.utils.KToast;
import com.kit.utils.Logger;

/**
 * @author: BaiCQ
 * @ClassName: BaseFragment
 * @date: 2018/8/17
 * @Description: Fragment 的基类
 */
public abstract class BaseFragment<P extends BasePresenter> extends Fragment implements IBasis, IBaseView {
    protected final String TAG = this.getClass().getSimpleName();
    protected BaseActivity activity;
    private View layout;
    private boolean init = false;//init 和 onRefresh()的执行的先后问题
    public P present;
    private LoadTag mLoadTag;

    @Override
    public final void onAttach(Context context) {
        super.onAttach(context);
        activity = (BaseActivity) context;
        UIStack.getInstance().add(this);
        mLoadTag = new LoadTag(activity);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        UIStack.getInstance().remove(this);
        init = false;
        Logger.e(TAG, "onDetach");
    }

    public abstract P createPresent();

    @Deprecated
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        layout = inflater.inflate(setLayoutId(), null);
        Logger.e(TAG, "onCreateView");
        return layout;
    }

    @Override
    public final void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Logger.e(TAG, "onViewCreated");
        present = createPresent();
        init();
        initListener();
        init = true;
    }

    @Override
    public abstract int setLayoutId();

    @Override
    public abstract void init();

    public abstract void initListener();

    protected View getLayout() {
        return layout;
    }

    protected <T extends View> T getView(@IdRes int id) {
        return layout.findViewById(id);
    }

    /**
     * 首次刷新尽量先于init执行
     */
    @Override
    public void onRefresh(Object obj) {
        Logger.e(TAG, "onRefresh");
    }

    @Override
    public void onNetChange() {
        Logger.e(TAG, "onRefresh");
    }

    public boolean isInit() {
        return init;
    }

    @Override
    public void showLoading(String msg) {
        mLoadTag.show(msg);
    }

    @Override
    public void dismissLoading() {
        mLoadTag.dismiss();
    }

    @Override
    public void showEmpty() {

    }

    @Override
    public void showToast(String message) {
        KToast.show(message);
    }
}

File: app/src/main/java/cn/rongcloud/voiceroomdemo/mvp/fragment/BaseFragment.kt
/*
 * Copyright © 2021 RongCloud. All rights reserved.
 */

package cn.rongcloud.voiceroomdemo.mvp.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.fragment.app.Fragment
import cn.rongcloud.voiceroomdemo.common.BaseLifeCyclePresenter
import cn.rongcloud.voiceroomdemo.common.IBaseView

/**
 * @author gusd
 * @Date 2021/06/24
 */
abstract class BaseFragment<P : BaseLifeCyclePresenter<V>, V : IBaseView>(@LayoutRes val layoutId: Int) :
    Fragment() {
    protected lateinit var presenter: P
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        presenter = initPresenter()
        lifecycle.addObserver(presenter)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return LayoutInflater.from(context).inflate(layoutId, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initView()
        initListener()
        initData()
    }

    open fun initData() {

    }

    open fun initListener() {

    }

    abstract fun initPresenter(): P

    abstract fun initView()

    open fun getTitle():String{
        return ""
    }
}